### ISSUE TITLE
Expand support for SDL_messagebox

### ISSUE THREAD

**Issue by flibitijibibo** [libsdl-org/SDL#6422]
This is a follow-up to #4227, which was originally just for Wayland.

While most platforms have a system for basic pop-ups, there are some that either A: don't have a messagebox system at all, or B: only have a basic messagebox system (i.e. you only get a body string and an OK button).

It's come up before, but as a 3.0 task: How should we accommodate systems with simpler messagebox APIs? We have a couple options...

1. Simplify the messagebox API in such a way that there is "simple" and "advanced" support, the downside being that not every platform will support advanced messageboxes
2. Make a multiplatform messagebox implementation using something like SDL_Render, the downside being that it would probably look horrible and would also need a font and text rendering support (a bonus upside would be controller support?)

Personally I'd lean towards option 1, but option 2 has come up enough times that maybe someone already has this buried somewhere.

**Comment by icculus** [comment]
> Make a multiplatform messagebox implementation using something like SDL_Render, the downside being that it would probably look horrible and would also need a font and text rendering support (a bonus upside would be controller support?)

It would also need to deal with figuring out what to do with existing rendering (for example, if you spin an SDL renderer on the Switch, it'll fail, since the only thing that is available there is the GL renderer, and if there's already a GL context, the Switch refuses to make a second.

I just think it would be messy, short of an extremely simply callback mechanism into the game that says "draw these pixels to the screen" and lets it deal with it, and hope that doing so doesn't trigger an SDL_assert on top of the one that generated the message box.

tl;dr: this isn't impossible, but it's messy enough that it isn't worth attempting, imho.

**Comment by icculus** [comment]
As for option 1: we already have a simple message box API, and honestly it's what 95% of things should use, so maybe just make the warning stronger on the not-simple API that it can fail.

**Comment by flibitijibibo** [comment]
> maybe just make the warning stronger on the not-simple API that it can fail.

That was my thought too - I think I've used the advanced system maybe like... once, ever?

> I just think it would be messy, short of an extremely simply callback mechanism into the game that says "draw these pixels to the screen" and lets it deal with it, and hope that doing so doesn't trigger an SDL_assert on top of the one that generated the message box.

This I hadn't thought of - we have something similar with FNA games: instead of telling people to use a messagebox we tell them to make a super basic window that barfs out the exception text and nothing else; maybe we could do an ImGui: Ultra Babby Edition just for messageboxes?

**Comment by slouken** [comment]
Hmm, that seems like a good idea, especially in the context of other game engines.

**Comment by ethindp** [comment]
This raises accessibility concerns. The normal messagebox APIs that most platforms expose have native accessibility support, so nothing need be done in that regard. However, if you tell the client application to do the drawing or use something like ImGui, you now need to implement accessibility yourself. This could probably trivially be done via something like AccessKit, but it's something to be aware of.

**Comment by icculus** [comment]
Also, if you're firing off a message box where you're willing to implement a callback into your renderer, you might as well just not use SDL_ShowMessageBox at all and roll your own basic UI that matches your game.

I know this isn't universally true, but message boxes are meant to be shown in disaster scenarios in most games, like, the game data files are missing so we don't even have a font to explain in the game window why we're about to terminate.

Also, the PlayStation and Switch SDKs both have an extremely basic facility built-in for this and you will absolutely fail cert if you ever use it in a release build.

As for the API...SDL_ShowSimpleMessageBox fits the bill well enough; it can probably ignore everything but the message text (flags, title, parent window) and still be Good Enough on a platform that can offer almost nothing at all.

Which is to say: maybe we just close this as WONTFIX?

**Comment by slouken** [comment]
Sounds good to me!
